src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -259,6 +259,141 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
             }
         }
 
+        /// <inheritdoc />
+        public override async Task CopyToAsync(PipeWriter destination, CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            CancellationTokenSource tokenSource = InternalTokenSource;
+            if (tokenSource.IsCancellationRequested)
+            {
+                ThrowHelper.ThrowOperationCanceledException_ReadCanceled();
+            }
+
+            if (_bufferedBytes > 0 && (!_examinedEverything || _isStreamCompleted))
+            {
+                BufferSegment? segment = _readHead;
+                try
+                {
+                    while (segment != null)
```

Comment:
I think we should enumerable until segment = _readTail

Reply:
When a `StreamPipeReader` hosts one single buffered segment then [`_readHead == _readTail`](https://github.com/dotnet/runtime/blob/d2fba8fdc44f5ed7a4676c219c626e36a3fa2998/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs#L317), we must loop until we get `null` as done in the [`Complete`](https://github.com/dotnet/runtime/blob/d2fba8fdc44f5ed7a4676c219c626e36a3fa2998/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs#L176) method.
I have added a test case to cover this situation.

Reply:
Yea, that makes sense.